flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/BigQueryDynamicTableSource.java

```diff
@@ -0,0 +1,233 @@
+                        .setColumnNames(DataType.getFieldNames(producedDataType))
+                        .build();
+    }
+
+    @Override
+    public void applyLimit(long limit) {
+        this.limit = (int) limit;
+    }
+
+    @Override
+    public Result applyFilters(List<ResolvedExpression> filters) {
+        Map<Boolean, List<Tuple3<Boolean, String, ResolvedExpression>>> translatedFilters =
+                filters.stream()
+                        .map(
+                                exp ->
+                                        Tuple2.<ResolvedExpression, Optional<String>>of(
+                                                exp, BigQueryRestriction.convert(exp)))
+                        .map(
+                                transExp ->
+                                        Tuple3.<Boolean, String, ResolvedExpression>of(
+                                                transExp.f1.isPresent(),
+                                                transExp.f1.orElse(""),
+                                                transExp.f0))
+                        .collect(
+                                Collectors.groupingBy(
+                                        (Tuple3<Boolean, String, ResolvedExpression> t) -> t.f0));
+        String rowRestriction =
+                translatedFilters.getOrDefault(true, Lists.newArrayList()).stream()
+                        .map(t -> t.f1)
+                        .collect(Collectors.joining(" AND "));
+        this.readOptions = this.readOptions.toBuilder().setRowRestriction(rowRestriction).build();
+        return Result.of(
+                translatedFilters.getOrDefault(true, Lists.newArrayList()).stream()
+                        .map(t -> t.f2)
+                        .collect(Collectors.toList()),
+                filters);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.readOptions, this.producedDataType, this.limit);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        final BigQueryDynamicTableSource other = (BigQueryDynamicTableSource) obj;
+        return Objects.equals(this.readOptions, other.readOptions)
+                && Objects.equals(this.producedDataType, other.producedDataType)
+                && Objects.equals(this.limit, other.limit);
+    }
+
+    @Override
+    public Optional<List<Map<String, String>>> listPartitions() {
+        BigQueryConnectOptions connectOptions = readOptions.getBigQueryConnectOptions();
+        BigQueryServices.QueryDataClient dataClient =
+                BigQueryServicesFactory.instance(connectOptions).queryClient();
+        return dataClient
+                // get the column name that is a partition, maybe none.
+                .retrievePartitionColumnName(
```

Comment:
[Good to have]

This is a very expensive way to get the partitioning information.
`retrievePartitionColumnName` runs a query to do this. Instead, it can be done using getTable which gives TableInfo.
And then use something like [this](https://github.com/GoogleCloudDataproc/spark-bigquery-connector/blob/0d5a980fa25aef61b91da7193cc8a25dabbecdf2/bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java#L495)

Can add a TODO here.

Reply:
Done, added the TODO on the code.

Reply:
BTW, this is addressed in #49